### ISSUE TITLE
go/types: vague docs about "incrementally" type checking files

### ISSUE THREAD

**Issue by mdempsky** [golang/go#20124]
The go/types documentation twice mentions "incrementally":

> Alternatively, create a new type checker with NewChecker and invoke it incrementally by calling Checker.Files.

> NewChecker returns a new Checker instance for a given package. Package files may be added incrementally via checker.Files.

However, it's unclear to me what that's supposed to mean. The types tutorial doesn't mention "incrementally" at all either, or even describe using the types.Checker API.

Intuitively, I would think an "incremental" API means the caller only needs to supply new information; e.g., to call Files({A}) and then Files({B}). But looking at the implementation, it appears the second call actually needs to be Files({A,B}). That's fine (and can't change now anyway), but I'm not clear in what sense it's "incremental." It doesn't build on any previous work, except to reuse the same maps previously allocated.

/cc @griesemer @alandonovan

**Comment by mateusz834** [comment]
> to call Files({A}) and then Files({B}). But looking at the implementation, it appears the second call actually needs to be Files({A,B})

Does this API even work for this?

```go
func TestTest(t *testing.T) {
	cfg := types.Config{
		Error: func(err error) {
			t.Log(err)
		},
	}

	fset := token.NewFileSet()
	c := types.NewChecker(&cfg, fset, types.NewPackage("", ""), nil)
	c.Files([]*ast.File{})

	const src = `package test; func test() {}`
	f, err := parser.ParseFile(fset, "file.go", src, parser.SkipObjectResolution|parser.ParseComments)
	if err != nil {
		t.Fatal(err)
	}

	const src2 = `package test; func test2() {}`
	f2, err := parser.ParseFile(fset, "file2.go", src2, parser.SkipObjectResolution|parser.ParseComments)
	if err != nil {
		t.Fatal(err)
	}

	t.Log(c.Files([]*ast.File{f}))     // <nil>
	t.Log(c.Files([]*ast.File{f, f2})) // test redeclared in this block
}
```

```sh
[mateusz@arch a ]$ go test . -v
=== RUN   TestTest
    main_test.go:34: <nil>
    main_test.go:14: file.go:1:20: test redeclared in this block
    main_test.go:14: file.go:1:20:      other declaration of test
    main_test.go:35: file.go:1:20: test redeclared in this block
--- PASS: TestTest (0.00s)
```

**Comment by adonovan** [comment]
> The go/types documentation twice mentions "incrementally":
> However, it's unclear to me what that's supposed to mean.

This idea was a misguided attempt to avoid the need for golang.org/x/tools/go/loader to type-check all the files twice: once as an importable package, then again, with additional *_test.go files, as an in-package test. The problem is that the new _test.go files can change the interpretation of the other files: for example, they can add declare new methods, causing existing x.f selectors to resolve differently; or they can add shadowing declarations for built-ins like any, max, or error. So there is no sound way to avoid type checking twice.

We should discourage anyone from trying to do it.

**Comment by mateusz834** [comment]
Yep, I somehow figured it out eventually that it seems to work if the current files do not depend on future ones. Thanks for the explanation. 

> We should discourage anyone from trying to do it.

Yep, the first line in package docs even references this, which is bad.

Maybe it should be even deprecated??
